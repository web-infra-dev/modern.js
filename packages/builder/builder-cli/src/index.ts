export { defineConfig } from '@modern-js/builder/cli';
