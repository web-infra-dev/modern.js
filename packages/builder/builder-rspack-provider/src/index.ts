export { getRspackVersion } from './shared/rspackVersion';
export { builderRspackProvider } from './provider';
export type { BuilderRspackProvider } from './provider';

export type {
  // Config Types
  BuilderConfig,
  NormalizedConfig,

  // Hook Callback Types
  ModifyRspackConfigFn,

  // Plugin Types
  BuilderPluginAPI,

  // Rspack
  Rspack,
  RspackConfig,
} from './types';
