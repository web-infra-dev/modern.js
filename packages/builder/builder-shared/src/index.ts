export * from './applyDefaultBuilderOptions';
export * from './constants';
export * from './createHook';
export * from './pluginStore';
export * from './createContext';
export * from './utils';
export * from './fs';
export * from './generateMetaTags';
export * from './getBrowserslist';
export * from './getCssSupport';
export * from './logger';
export * from './mergeBuilderConfig';
export * from './onExitProcess';
export * from './pick';
export * from './regexp';
export * from './types';
export * from './config';
export * from './format';
export * from './devServer';
export * from './prodServer';
export * from './chain';
export * from './schema';
export * from './plugins';
export * from './openBrowser';
export * from './apply';
export * from './fallback';
export * from './getLoaderOptions';
export * from './svgo';
export * from './patch';
export * from './manifest';
export * from './css';
export * from './minimize';
export * from './core-js';
export * from './progress';
export * from './prettyTime';
