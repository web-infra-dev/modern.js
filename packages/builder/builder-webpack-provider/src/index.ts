export { builderWebpackProvider } from './provider';
export type { BuilderWebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export type { WebpackBuildError } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  BuilderPlugin,
  BuilderPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CSSExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
