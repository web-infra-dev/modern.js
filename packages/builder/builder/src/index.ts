export { createBuilder } from './createBuilder';
export { mergeBuilderConfig } from '@modern-js/builder-shared';

export { builderPluginSourceBuild } from './plugins/sourceBuild';

export type {
  BuilderMode,
  BuilderEntry,
  BuilderTarget,
  BuilderPlugin,
  BuilderContext,
  BuilderInstance,
  CreateBuilderOptions,
  InspectConfigOptions,

  // Hook Callback Types
  OnExitFn,
  OnAfterBuildFn,
  OnAfterCreateCompilerFn,
  OnAfterStartDevServerFn,
  OnBeforeBuildFn,
  OnBeforeStartDevServerFn,
  OnBeforeCreateCompilerFn,
  OnDevCompileDoneFn,
  ModifyBuilderConfigFn,
} from '@modern-js/builder-shared';
