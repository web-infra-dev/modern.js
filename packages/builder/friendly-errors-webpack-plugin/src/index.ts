export * from './shared/types';
export * from './shared/utils';
export * from './plugin';
