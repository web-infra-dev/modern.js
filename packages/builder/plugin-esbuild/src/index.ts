import {
  JS_REGEX,
  TS_REGEX,
  applyScriptCondition,
} from '@modern-js/builder-shared';
import type { BuilderPlugin } from '@modern-js/builder';
import type { BuilderPluginAPI } from '@modern-js/builder-webpack-provider';
import type {
  LoaderOptions,
  MinifyPluginOptions,
} from '../compiled/esbuild-loader/interfaces';

export interface PluginEsbuildOptions {
  loader?: false | LoaderOptions;
  minimize?: false | MinifyPluginOptions;
}

export function builderPluginEsbuild(
  userOptions: PluginEsbuildOptions = {},
): BuilderPlugin<BuilderPluginAPI> {
  return {
    name: 'builder-plugin-esbuild',

    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, isProd, target }) => {
        const builderConfig = api.getNormalizedConfig();
        const compiledEsbuildLoaderPath = require.resolve(
          '../compiled/esbuild-loader',
        );

        const options: PluginEsbuildOptions = {
          loader: {
            target: 'es2015',
            charset: builderConfig.output.charset,
          },
          minimize: {
            css: true,
            target: 'es2015',
            format: target === 'web' ? 'iife' : undefined,
          },
          ...userOptions,
        };

        if (options.loader !== false) {
          // remove babel-loader and ts-loader
          chain.module.rule(CHAIN_ID.RULE.JS).uses.delete(CHAIN_ID.USE.BABEL);
          chain.module
            .rule(CHAIN_ID.RULE.TS)
            .uses.delete(CHAIN_ID.USE.BABEL)
            .delete(CHAIN_ID.USE.TS);

          // add esbuild-loader
          chain.module
            .rule(CHAIN_ID.RULE.JS)
            .test(JS_REGEX)
            .use(CHAIN_ID.USE.ESBUILD)
            .loader(compiledEsbuildLoaderPath)
            .options({
              loader: 'jsx',
              ...options?.loader,
            });

          const rule = chain.module.rule(CHAIN_ID.RULE.TS);
          rule
            .test(TS_REGEX)
            .use(CHAIN_ID.USE.ESBUILD)
            .loader(compiledEsbuildLoaderPath)
            .options({
              loader: 'tsx',
              ...options?.loader,
            });
          applyScriptCondition({
            rule,
            config: builderConfig,
            context: api.context,
            includes: [],
            excludes: [],
          });
        }

        if (isProd && options.minimize !== false) {
          const { ESBuildMinifyPlugin } = await import(
            '../compiled/esbuild-loader'
          );

          // @ts-expect-error webpack-chain missing minimizers type
          chain.optimization.minimizers
            .delete(CHAIN_ID.MINIMIZER.JS)
            .delete(CHAIN_ID.MINIMIZER.CSS);

          chain.optimization
            .minimizer(CHAIN_ID.MINIMIZER.ESBUILD)
            .use(ESBuildMinifyPlugin)
            .init(
              () =>
                new ESBuildMinifyPlugin({
                  // other legalComments such as linked is not supported yet
                  // https://github.com/privatenumber/esbuild-loader/issues/263
                  legalComments:
                    builderConfig.output?.legalComments === 'none'
                      ? 'none'
                      : 'inline',
                  ...options?.minimize,
                }),
            );
        }
      });
    },
  };
}

/**
 * @deprecated Using builderPluginEsbuild instead.
 */
export const PluginEsbuild = builderPluginEsbuild;
