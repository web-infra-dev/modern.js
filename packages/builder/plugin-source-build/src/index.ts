import { setConfig } from '@modern-js/builder-shared';
import type { BuilderPlugin } from '@modern-js/builder';
import type { BuilderPluginAPI as WebpackBuilderPluginAPI } from '@modern-js/builder-webpack-provider';
import type { BuilderPluginAPI as RspackBuilderPluginAPI } from '@modern-js/builder-rspack-provider';
import {
  type ExtraMonorepoStrategies,
  getDependentProjects,
  filterByField,
} from '@modern-js/monorepo-utils';

export const pluginName = 'builder-plugin-source-build';

export const getSourceInclude = async (options: {
  projectNameOrRootPath: string;
  findMonorepoStartPath: string;
  sourceField: string;
  extraMonorepoStrategies?: ExtraMonorepoStrategies;
}) => {
  const {
    projectNameOrRootPath,
    sourceField,
    extraMonorepoStrategies,
    findMonorepoStartPath,
  } = options;
  const projects = await getDependentProjects(projectNameOrRootPath, {
    cwd: findMonorepoStartPath,
    recursive: true,
    filter: filterByField(sourceField),
    extraMonorepoStrategies,
  });

  const includes = [];
  for (const project of projects) {
    includes.push(...project.getSourceEntryPaths());
  }

  return includes;
};

export interface PluginSourceBuildOptions {
  projectName?: string;
  sourceField?: string;
  extraMonorepoStrategies?: ExtraMonorepoStrategies;
}

/**
 * Usage:
 *
 *   const { builderPluginSourceBuild } = await import('@modern-js/builder-plugin-source-build');
 *
 *   builder.addPlugins([ builderPluginSourceBuild() ]);
 */
export function builderPluginSourceBuild(
  options?: PluginSourceBuildOptions,
): BuilderPlugin<WebpackBuilderPluginAPI | RspackBuilderPluginAPI> {
  const {
    projectName,
    sourceField = 'source',
    extraMonorepoStrategies,
  } = options ?? {};
  return {
    name: pluginName,

    async setup(api) {
      const projectRootPath = api.context.rootPath;
      api.modifyBuilderConfig(async config => {
        const includes = await getSourceInclude({
          projectNameOrRootPath: projectName ?? projectRootPath,
          sourceField,
          findMonorepoStartPath: projectRootPath,
          extraMonorepoStrategies,
        });
        config.source = config.source ?? {};
        config.source.include = [...(config.source.include ?? []), ...includes];
      });

      api.modifyBundlerChain(chain => {
        // Now not support chain.resolve.conditionNames API
        // chain.resolve.conditionNames.prepend(sourcePkgField);

        // when user not config source.resolveMainFields, mainFields is empty array
        if (chain.resolve.mainFields.values().length === 0) {
          // "..." is special syntax,it will retain the original value
          chain.resolve.mainFields.prepend('...');
          chain.resolve.mainFields.prepend(sourceField);
        }
      });

      // when support chain.resolve.conditionNames API, remove this logic
      if (api.context.bundlerType === 'rspack') {
        (api as RspackBuilderPluginAPI).modifyRspackConfig(async config => {
          setConfig(config, 'resolve.conditionNames', [
            '...', // Special syntax: retain the original value
            sourceField,
            ...(config.resolve?.conditionNames ?? []),
          ]);
        });
      } else {
        (api as WebpackBuilderPluginAPI).modifyWebpackConfig(async config => {
          config.resolve = config.resolve ?? {};
          config.resolve.conditionNames = [
            '...', // Special syntax: retain the original value
            sourceField,
            ...(config.resolve.conditionNames ?? []),
          ];
        });
      }
    },
  };
}
