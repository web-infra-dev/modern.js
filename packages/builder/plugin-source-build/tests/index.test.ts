import { expect, describe, it } from 'vitest';
import { createStubBuilder } from '@modern-js/builder-webpack-provider/stub';
import { builderPluginSourceBuild } from '../src';

describe('plugins/source-build', () => {
  it('should add source-build config', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginSourceBuild()],
    });
    const config = await builder.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
