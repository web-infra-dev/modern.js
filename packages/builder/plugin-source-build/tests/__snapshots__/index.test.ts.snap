// Vitest Snapshot v1

exports[`plugins/source-build > should add source-build config 1`] = `
{
  "resolve": {
    "conditionNames": [
      "...",
      "source",
    ],
    "mainFields": [
      "source",
      "...",
    ],
  },
}
`;
